agent-inject/agent/agent.go

```diff
@@ -355,6 +368,67 @@ func ShouldInject(pod *corev1.Pod) (bool, error) {
 	return true, nil
 }
 
+// ShouldDelete checks whether the pod in question has already been injected
+// and the generated config map needs to be deleted.
+func ShouldDelete(pod *corev1.Pod) (bool, error) {
+	raw, ok := pod.Annotations[AnnotationAgentInject]
+	if !ok {
+		return false, nil
+	}
+
+	inject, err := strconv.ParseBool(raw)
+	if err != nil {
+		return false, err
+	}
+
+	if !inject {
+		return false, nil
+	}
+
+	// This shouldn't happen so bail.
+	raw, ok = pod.Annotations[AnnotationAgentStatus]
+	if !ok {
+		return false, nil
+	}
+
+	// Previously injected agents
+	if raw != "injected" {
+		return false, nil
+	}
+
+	// Only delete config maps if we made one (ignore custom user defined maps)
+	raw, ok = pod.Annotations[AnnotationAgentGeneratedConfigMapName]
+	if !ok {
+		return false, nil
+	}
+
+	if raw == "" {
+		return false, nil
+	}
+
+	// Only delete config maps if the name of the configmap and the pod are the same.
+	if ConfigMapName(pod) != pod.Annotations[AnnotationAgentGeneratedConfigMapName] {
```

Comment:
Under what conditions would be expect to take this path in the logic? If the user has modified the annotation themselves? IIUC it's quite a confusing state to be in, so perhaps it's worth logging something to say "something weird has happened so we'll do the safe thing and leave it alone - note that modifying the agent-configmap-name annotation will cause dangling config maps like this to be left around. The config map with <this> name is _probably_ safe to delete, but check it yourself".

Reply:
@tvoran and I agreed we needed to protect against someone trying to delete a ConfigMap owned by another pod. You could craft a pod name, try a bogus injection and then delete it to trigger a privileged delete on the configmap. This check just ensures this can't be abused.